share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorService.java

```diff
@@ -0,0 +1,533 @@
+                    "Share coordinator is not available."
+                )
+            );
+        }
+
+        // The request received here could have multiple keys of structure group:topic:partition. However,
+        // the readState method in ShareCoordinatorShard expects a single key in the request. Hence, we will
+        // be looping over the keys below and constructing new ReadShareGroupStateRequestData objects to pass
+        // onto the shard method.
+
+        request.topics().forEach(topicData -> {
+            Uuid topicId = topicData.topicId();
+            topicData.partitions().forEach(partitionData -> {
+                // Request object containing information of a single topic partition
+                ReadShareGroupStateRequestData requestForCurrentPartition = new ReadShareGroupStateRequestData()
+                    .setGroupId(groupId)
+                    .setTopics(Collections.singletonList(new ReadShareGroupStateRequestData.ReadStateData()
+                        .setTopicId(topicId)
+                        .setPartitions(Collections.singletonList(partitionData))));
+                SharePartitionKey coordinatorKey = SharePartitionKey.getInstance(request.groupId(), topicId, partitionData.partition());
+                // Scheduling a runtime read operation to read share partition state from the coordinator in memory state
+                CompletableFuture<ReadShareGroupStateResponseData> future = runtime.scheduleReadOperation(
+                    "read-share-group-state",
+                    topicPartitionFor(coordinatorKey),
+                    (coordinator, offset) -> coordinator.readState(requestForCurrentPartition, offset)
+                ).exceptionally(exception -> handleOperationException(
+                    "read-share-group-state",
+                    request,
+                    exception,
+                    (error, message) -> ReadShareGroupStateResponse.toErrorResponseData(
+                        topicData.topicId(),
+                        partitionData.partition(),
+                        error,
+                        "Unable to read share group state: " + exception.getMessage()
+                    ),
+                    log
+                ));
+                futureMap.computeIfAbsent(topicId, k -> new HashMap<>());
+                futureMap.get(topicId).put(partitionData.partition(), future);
+            });
+        });
+
+        // Combine all futures into a single CompletableFuture<Void>
+        CompletableFuture<Void> combinedFuture = CompletableFuture.allOf(futureMap.values().stream()
+            .flatMap(map -> map.values().stream()).toArray(CompletableFuture[]::new));
+
+        // Transform the combined CompletableFuture<Void> into CompletableFuture<ReadShareGroupStateResponseData>
+        return combinedFuture.thenApply(v -> {
+            List<ReadShareGroupStateResponseData.ReadStateResult> readStateResult = new LinkedList<>();
```

Comment:
Why a linked list here (and other similar places)? Seems like we know the expected size of the resulting list, so we could allocate an ArrayList with the size.